MVP of Game of Life

fedacking opened the pull request:
Basic Implementation of Game of Life, with a graphical representation and no user interaction.

fedacking committed:
Added makefile and README

Adding a basic Makefile and README to provide
explanation and build structure to the repo.

fedacking committed:
Merge pull request #1 from fedacking/feature/README

Added makefile and README

fedacking committed:
Added CI-CD through Github Workflows

Added basic CI-CD job that runs a format,
clippy and the tests for the project.

fedacking committed:
Merge pull request #2 from fedacking/feature/ci-cd

Added CI-CD through Github Workflows

fedacking committed:
Initial Logic commit

Added the basic logic of the game of life, with a
structure for the data, and an update function.
All of that is present in the gol.rs file, the
main file just has some basic test code.

fedacking committed:
Added test

Added multiple tests checking the basic rules of the Game of Life

fedacking committed:
Fixed CI-CD

Changed it to ci-cd to do an all for the format and use the makefile

fedacking committed:
Update README.md

fedacking committed:
Fixes from comments

- Added a println to divide the gol boards in the main program
- Used more comparct formatting instead of nested matches for readability.

fedacking committed:
Merge pull request #3 from fedacking/feature/gol-logic

Initial Logic commit

fedacking committed:
Fixed missing make build for CI-CD config

fedacking committed:
Added link and formating to README

fedacking committed:
Merge pull request #5 from fedacking/fix/make-ci-cd

Fix/make ci-cd, by adding a more complete format, a better description of the problem and expanding the makefile.

fedacking committed:
Fixed update_cell range limit

Fixed a bug where the range was improperly set
for the update_cell, and added a new test to
catch the mistake.

fedacking committed:
Cleaner range

fedacking committed:
Merge pull request #6 from fedacking/fix/fixed-logic

Fixed update_cell range limit

fedacking committed:
UI

Added a ui through macroquad to view the game of life evolve

fedacking committed:
Added comments and tests

fedacking committed:
Merge pull request #7 from fedacking/feature/ui

Feature/UI

fedacking committed:
Misc Format changes

- Removed unused function
- Ran Make Format
- Aadded trailing newlines to the workflow

fedacking committed:
Added newline

fedacking committed:
Merge pull request #8 from fedacking/fix/improved_transition_tests

Fix/improved transition tests

gabrielbosio commented on README.md:
```diff
@@ -1,2 +1,45 @@
+
+### Constraints
+
+Because we have limited memory, our game of life is limited to the a square of size defined at compile time in our main.rs source file. 
```
I would add something like `This implementation stores the Game of Life universe completely on memory` at the beginning to connect memory limitation with the data structure that represents the universe.

gabrielbosio replied:
Also:

```suggestion
Because we have limited memory, our game of life is limited to a square of size defined at compile time in our main.rs source file. 
```

gabrielbosio commented on src/gol.rs:
```diff
@@ -0,0 +1,213 @@
+    const A: CellState = CellState::Alive;
+
+    #[test]
+    // Verifies that the empty pattern should remove patterns
```
```suggestion
    // Verifies that the empty pattern does not change the cell states of the board
```

gabrielbosio commented on README.md:
```diff
@@ -1,2 +1,45 @@
+Because we have limited memory, our game of life is limited to the a square of size defined at compile time in our main.rs source file. 
+
+The Game of life comes with a graphical user interface that updates at the rate of 2 times per second. This and the size of the cells on screen can be changed through consts in main.rs (respectively UPS and CELL_SIZE).
+
```
I would add here that, for now, the initial state is fixed and to change it, you need to recompile the crate.

fedacking committed:
Fixed Readme and Documentation

juanbono commented on src/gol.rs:
```diff
@@ -0,0 +1,213 @@
+        indexes
+    }
+
+    pub fn count_neighbours(&self, x: usize, y: usize) -> usize {
```
can you add some docs to this method? In general it's a good idea to add docs to public data structures/functions

fedacking replied:
It should have been private, but I will add it, maybe it's a useful function for someone?

juanbono commented on src/main.rs:
```diff
@@ -0,0 +1,98 @@
+    draw_line(x, y, x, y + CELL_SIZE, 2.0, BLACK);
+}
+
+// Traverses the board to draw it. To see the details of how the board is drawn,
```
add an extra `/` to convert it to documentation visible by the editor.

juanbono commented on src/main.rs:
```diff
@@ -0,0 +1,98 @@
+    // If you want to change the size it's necessary to also change the
+    // Contents, as it expects an array of the correct size
+    // Or change to a ::new and set the specific changed cells by hand
+    /*let mut gol: GameOfLife<10, 10> = GameOfLife::from_board([
```
here you can add:

```rust  before the code to highlight its nature (this is useful for documentation purposes since it will show differently)

juanbono commented on src/gol.rs:
```diff
@@ -0,0 +1,213 @@
+use std::ops::{Index, IndexMut};
+
+#[derive(Debug, Clone, Copy, PartialEq)]
```
add some docs to this struct

juanbono commented on src/main.rs:
```diff
@@ -0,0 +1,98 @@
+
+#[macroquad::main("Fedacking's Game of Life")]
+async fn main() {
+    //const D: CellState = CellState::Dead;
```
remove commented code.

fedacking committed:
Fixes of documentation

Following Juanbono's fixes
- Removed superflous commented code line
- Added documentation to the CellState enum and the GameOfLife struct
- Added `to the commented code for rustdoc
- Converted documentation to rustdoc compatible format

fedacking committed:
Format

fedacking committed:
Fix missing word

DiegoCivi reviewed (approved):
Looks good!
